app-creator/src/widgets/ui-map/ui-map.ts

```diff
@@ -1,72 +1,230 @@
+   * https://developers.google.com/maps/documentation/javascript/reference/map#MapOptions.
+   */
+  @property({ type: Object }) mapOptions: google.maps.MapOptions = {};
 
   /**
    * Sets map element.
    */
   private map: google.maps.Map | null = null;
 
-  getMap() {
-    return this.map;
-  }
+  /**
+   * Adds edit icon to panel.
+   */
+  @property({ type: Boolean }) editable = false;
 
-  setMap(map: google.maps.Map | null) {
-    this.map = map;
+  connectedCallback() {
+    this.initMap();
   }
 
-  connectedCallback() {}
+  initMap() {
+    loadGoogleMaps(this.apiKey).then(() => {
+      this.mapOptions.zoom = this.zoom || 0;
+
+      this.mapOptions.center = this.center;
+
+      this.map = new google.maps.Map(this, this.mapOptions);
+      this.dispatchEvent(
+        new CustomEvent('google-map-ready', { detail: this.map })
+      );
+    });
+  }
 
   // Disable the shadow root, since it interferes with Google Maps.
   // https://lit-element.polymer-project.org/guide/templates#renderroot
   createRenderRoot() {
     return this;
   }
 
+  /**
+   * Triggered when the edit icon is clicked. Stores a reference of the selected element in the store and
+   * displays a set of inputs for editing its attributes.
+   */
+  handleEditWidget() {
+    // Check if a widgetRef has been set.
+    const ref = store.getState().template[this.id].widgetRef;
+    if (ref == null) {
+      // Set the ref to be the current element.
+      store.dispatch(updateWidgetRef(this));
+    }
+    store.dispatch(setEditingWidget(this));
+  }
+
   render() {
-    return html``;
-  }
-
-  init(): void {
-    const mapOptions = {
-      center: this.center,
-      zoom: this.zoom,
-      maxZoom: 10,
-      streetViewControl: false,
-    };
-
-    const mapContainer = document.createElement('div');
-    mapContainer.style.width = '100%';
-    mapContainer.style.height = '100%';
-    this.appendChild(mapContainer);
-    // Render the base Google Map directly into the component.
-    this.map = new google.maps.Map(mapContainer, mapOptions);
+    const { editable, handleEditWidget } = this;
+    const editableMarkup = editable
+      ? html`
+          <div id="editable-view">
+            <iron-icon
+              class="edit-buttons"
+              icon="create"
+              @click=${handleEditWidget}
+            ></iron-icon>
+          </div>
+        `
+      : nothing;
+    return html` ${editableMarkup}`;
+  }
+
+  setAttribute(key: string, value: any) {
+    switch (key) {
+      case 'zoom':
+        this.zoom = parseInt(value);
+        break;
+      case 'center':
+        this.center = value;
+        break;
+      case 'latitude':
+        this.center = { ...this.center, lat: parseFloat(value) };
+        break;
+      case 'longitude':
+        this.center = { ...this.center, lng: parseFloat(value) };
+        break;
+      case 'map':
+        this.map = value;
+        break;
+    }
+
+    this.initMap();
```

Comment:
If I understand correctly, the implication is that when any of the properties are set and the element rerenders, we have to reinitialize the map since the underlying dom changes. We should be able to cache the maps dom element and inject it into the new DOM container (https://stackoverflow.com/questions/5042170/clone-google-map-instance), so that we can simply call `this.map.setZoom` in this function. This might explain why things start to slow down when we have maps on the screen, since the Maps API is constantly having to reinitialize itself into the rendered containers. What do you think?

Reply:
This is what we're doing now so we don't have to reinitialize the map: 

```Typescript
if (this.map == null) {
     this.map = new google.maps.Map(this, this.mapOptions);
} else {
     this.map.setOptions(this.mapOptions);
}
```